### PR TITLE
Add per-socket keepalive options

### PR DESCRIPTION
**Pull request by ywelsch**
Uses JDK 11's per-socket configuration of TCP keepalive (supported on Linux and Mac), see https://bugs.openjdk.java.net/browse/JDK-8194298, and exposes these as transport settings. By default, these options are disabled for now (i.e. fall-back to OS behavior), but I would like to explore whether we can enable them by default, in particular to force keepalive configurations that are better tuned for running ES.

### CODE REVIEW

**Comment by original-brownbear** [comment]
@ywelsch this is failing third party tests:

```
10:27:39 > Task :libs:elasticsearch-nio:forbiddenApisMain FAILED
10:27:39 Forbidden class/interface use: jdk.net.ExtendedSocketOptions [non-portable or internal runtime class]
10:27:39   in org.elasticsearch.nio.ChannelFactory$RawChannelFactory (ChannelFactory.java:226)
10:27:39 Forbidden class/interface use: jdk.net.ExtendedSocketOptions [non-portable or internal runtime class]
10:27:39   in org.elasticsearch.nio.ChannelFactory$RawChannelFactory (ChannelFactory.java:227)
10:27:39 Forbidden class/interface use: jdk.net.ExtendedSocketOptions [non-portable or internal runtime class]
10:27:39   in org.elasticsearch.nio.ChannelFactory$RawChannelFactory (ChannelFactory.java:229)
10:27:39 Forbidden class/interface use: jdk.net.ExtendedSocketOptions [non-portable or internal runtime class]
10:27:39   in org.elasticsearch.nio.ChannelFactory$RawChannelFactory (ChannelFactory.java:230)
10:27:39 Forbidden class/interface use: jdk.net.ExtendedSocketOptions [non-portable or internal runtime class]
10:27:39   in org.elasticsearch.nio.ChannelFactory$RawChannelFactory (ChannelFactory.java:232)
10:27:39 Forbidden class/interface use: jdk.net.ExtendedSocketOptions [non-portable or internal runtime class]
10:27:39   in org.elasticsearch.nio.ChannelFactory$RawChannelFactory (ChannelFactory.java:233)
10:27:39 Scanned 28 class file(s) for forbidden API invocations (in 0.23s), 6 error(s).
10:27:39 
```

File: docs/reference/modules/transport.asciidoc [hunk 1]
```diff
@@ -84,6 +84,15 @@ example above:
 * `publish_host`: The host which is published in informational APIs
 * `tcp.no_delay`: Configures the `TCP_NO_DELAY` option for this socket
 * `tcp.keep_alive`: Configures the `SO_KEEPALIVE` option for this socket
+* `tcp.keep_idle`: Configures the `TCP_KEEPIDLE` option for this socket.
```
Comment (DaveCTurner) [inline docs/reference/modules/transport.asciidoc]:
```suggestion
* `tcp.keep_idle`: Configures the `TCP_KEEPIDLE` option for this socket, which determines the time in seconds that incoming connections must be idle before starting to send TCP keepalive probes.
```

File: docs/reference/modules/transport.asciidoc [hunk 2]
```diff
@@ -84,6 +84,15 @@ example above:
...
+   Only applicable on Linux and Mac. Defaults to -1, which does not set this
+   option at the socket level, but uses default system configuration instead.
+* `tcp.keep_interval`: Configures the `TCP_KEEPINTVL` option for this socket.
```
Comment (DaveCTurner) [inline docs/reference/modules/transport.asciidoc]:
```suggestion
* `tcp.keep_interval`: Configures the `TCP_KEEPINTVL` option for this socket, which determines the time in seconds between sending TCP keepalive probes.
```

File: docs/reference/modules/transport.asciidoc [hunk 3]
```diff
@@ -84,6 +84,15 @@ example above:
...
+   Only applicable on Linux and Mac. Defaults to -1, which does not set this
+   option at the socket level, but uses default system configuration instead.
+* `tcp.keep_count`: Configures the `TCP_KEEPCNT` option for this socket.
```
Comment (DaveCTurner) [inline docs/reference/modules/transport.asciidoc]:
```suggestion
* `tcp.keep_count`: Configures the `TCP_KEEPCNT` option for this socket, which determines the number of unacknowledged TCP keepalive probes that may be sent on a connection before it is dropped.
```

**Comment by ywelsch** [comment]
@elasticmachine test this please

**Comment by tbrooks8** [comment]
I started looking at this today and should be able to finishing reviewing it before Monday.

**Review by Tim-Brooks — APPROVED** [review]
LGTM

**Review by original-brownbear — APPROVED** [review]
LGTM2

**Comment by colings86** [comment]
It looks like the backporting is complete here so I have removed the backport pending label while generating 7.4.0 release notes. Please let me know if this is wrong

**Pull request merged**